drivers/net/ethernet/intel/idpf/idpf_aux.c

```diff
@@ -0,0 +1,327 @@
+	eth_dev->aux_info.caps.other_caps = adapter->caps.other_caps;
+	eth_dev->aux_info.caps.max_tx_hdr_size = adapter->caps.max_tx_hdr_size;
+	eth_dev->aux_info.caps.max_sg_bufs_per_tx_pkt =
+		adapter->caps.max_sg_bufs_per_tx_pkt;
+	eth_dev->aux_info.caps.min_sso_packet_len =
+		adapter->caps.min_sso_packet_len;
+}
+
+/**
+ * idpf_aux_release - Release IDPF aux device associated with auxiliary device
+ * @device: pointer to the device
+ */
+static void idpf_aux_release(struct device *device)
+{
+	struct auxiliary_device *adev = to_auxiliary_dev(device);
+	struct idpf_auxiliary_device *aux_dev;
+
+	aux_dev = container_of(adev, struct idpf_auxiliary_device,
+			       adev);
+	ida_free(&idpf_aux_dev_ida, aux_dev->adev.id);
+	kfree(aux_dev);
+}
+
+/**
+ * idpf_aux_alloc_dev - allocate one auxiliary device
+ * @adapter: IDPF main private structure
+ * @index: auxiliary device index witin the same function
+ * @vport_type: Vport type
+ *
+ * Returns allocated structure
+ */
+static struct idpf_auxiliary_device *
+idpf_aux_alloc_dev(struct idpf_adapter *adapter, unsigned int index,
+		   enum idpf_vport_type vport_type)
+{
+	struct idpf_auxiliary_device *aux_dev;
+	struct idpf_vport_max_q *max_q;
+	u16 err;
+
+	aux_dev = kzalloc(sizeof(*aux_dev), GFP_KERNEL);
+	if (!aux_dev)
+		return NULL;
+
+	aux_dev->adev.id = ida_alloc(&idpf_aux_dev_ida, GFP_KERNEL);
+	aux_dev->adev.name = "eth";
+	aux_dev->adev.dev.release = idpf_aux_release;
+	aux_dev->adev.dev.parent = &adapter->pdev->dev;
+
+	aux_dev->aux_info.aux_shared = &adapter->aux_shared;
+	aux_dev->aux_info.vport_type = vport_type;
+	aux_dev->aux_info.id = index;
+	aux_dev->aux_info.vport_id = -1;
+	max_q = &aux_dev->aux_info.caps.q_info;
+	err = idpf_vport_alloc_max_qs(adapter, max_q,
+				      vport_type == IDPF_DEFAULT_VPORT);
+	if (err) {
+		kfree(aux_dev);
+		return NULL;
+	}
+
+	return aux_dev;
+}
+
+/**
+ * idpf_aux_unplug_device - unplug one auxiliary device
+ * @adapter: IDPF main private structure
+ * @aux_dev: the auxiliary device to be removed from the auxiliary bus
+ */
+static void idpf_aux_unplug_device(struct idpf_adapter *adapter,
+				   struct idpf_auxiliary_device *aux_dev)
+{
+	struct idpf_vport_max_q *max_q;
+
+	max_q = &aux_dev->aux_info.caps.q_info;
+
+	idpf_vport_dealloc_max_qs(adapter, max_q);
+	auxiliary_device_delete(&aux_dev->adev);
+	auxiliary_device_uninit(&aux_dev->adev);
+}
+
+/**
+ * idpf_aux_plug_devices - Allocate, initialize and add auxiliary devices
+ * @adapter: IDPF main private structure
+ *
+ * Returns 0 in case of success, negative error number otherwise.
+ */
+int idpf_aux_plug_devices(struct idpf_adapter *adapter)
+{
+	struct device *dev = &adapter->pdev->dev;
+	struct idpf_auxiliary_device *aux_dev;
+	unsigned int i;
+	int err;
+
+	adapter->adevs =
+		kzalloc((adapter->max_vports *
+			 sizeof(struct idpf_auxiliary_device *)),
+			GFP_KERNEL);
+	if (!adapter->adevs)
+		return -ENOMEM;
+
+	for (i = 0; i < idpf_get_default_vports(adapter); i++) {
```

Comment:
Whole body of for can be separate function to plug one aux dev, rest of the code should be in pci driver, as not all consumer will do the same here (ixd for example). Also, I assume that more than default vport vports can be created, so separate function for adding one will be useful.